### PR TITLE
Refactor/ether swr guilds

### PR DESCRIPTION
**Pull request by Kenny-Gin1**
# Description

This is the third part of three to refactoring our hooks to use `wagmi` completely. With this third refactoring, we have finally sunset `ether-swr`. This will hopefully come together as a better experience for the end-user since we are now optimizing how we fetch data from our contracts. E.g. polling.

I added some tests to it as well.

- Refactoring and moving all the `hooks/Guilds/guild` to `Module/Guilds/Hooks`.
- Adding tests for all the hooks.

Note that the tests just mock out the entire file. It is not optimal - but the purpose of these tests is rather to enhance the developer experience and quickly understand what should be returned from the hooks.

NOTE: This is a very big refactor and moving of files. I have tried to check for as many errors and/or bugs as possible manually that might have appeared with this refactor. But please be vigilant of any changes that you might see.

Questions to ponder in this PR: 
- The test files basically mock out the entire corresponding hook file and the purpose is for them to understand the usage of the hooks and what it is meant to return. Is this useful or not? If not, I will remove all of them.
- There is one known error I am getting at the moment `Invalid BigNumber value argument="value", value=null, code=INVALID_ARGUMENT`. I cannot see that it breaks anything at the moment and am not sure if it was introduced in this branch or not.

Closes # ([issue](https://github.com/DXgovernance/DAVI/issues/278))

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Unit-testing and manual testing.

# Checklist:

- [ ] My code follows the existing style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any UI changes have been tested and made responsive for mobile views

### CODE REVIEW

**Comment by Filipv95** [comment]
@Kenny-Gin1 Hey, can you ping me when it's ready to test, also if you can point out what should be specifically tested in this PR, ty ^^

**Comment by Kenny-Gin1** [comment]
@Filipv95 This PR is now ready for testing. It is the biggest of the three refactors as of yet, so I would say that it requires extensive regression testing. Let me know if you see any weird behavior and/or information, and such. Thanks :)

**Comment by Filipv95** [comment]
Hey @Kenny-Gin1, I've finished regression testing, did'not notice anything unusually, couple of issues I've noticed are present on develop branch as well. 

tldr, looks good, approving. ^^

**Review by Filipv95 — APPROVED** [review]

File: src/hooks/Guilds/useUpdateEnsContent.ts [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
 
-/*
-  @from: Guild address
-  @to ENS resolver address
-  @node: Namehash of the ENS name 
-  @contentHash: IPFS/CID hash
-  @optionalProps: Props needed to update the ENS frontend content
-  */
+/**
+ * @property {string} from Guild address
```
Comment (mprasanjith) [inline src/hooks/Guilds/useUpdateEnsContent.ts:5]: Where are these comments used? Are we planning to setup JSDoc or something?

File: src/hooks/Guilds/erc20/useERC20Info.ts [hunk 2]
```diff
@@ -40,8 +40,10 @@ export const useERC20Info = (contractAddress: string): ERC20InfoHook => {
       },
     ],
   });
-  if (!data || data.every(v => v === null))
+  if (!data || data.every(v => v === null)) {
+    console.log('useERC20Info', data);
```
Comment (mprasanjith) [inline src/hooks/Guilds/erc20/useERC20Info.ts:44]: I think we can get rid of the console.log

File: src/bytecodes/local.json [hunk 3]
```diff
@@ -1,27 +1,27 @@
 [
   {
     "type": "ERC20Guild",
-    "bytecode_hash": "0x083340048d4df973b5de9747b5f63894c40f0e83dea83d60f976be457a69674e",
+    "bytecode_hash": "0x23573d1da41662c45c638884d6b44805ffcd2c3423cd5b30e6ab41367296cbad",
```
Comment (mprasanjith) [inline src/bytecodes/local.json:4]: Not sure if these hashes are the latest.

File: src/Modules/Guilds/Hooks/useTotalLocked.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  if (isSnapshotGuild) {
+    return {
+      data: totalLockedAtProposalSnapshotResponse
+        ? BigNumber.from(totalLockedAtProposalSnapshotResponse)
+        : undefined,
+    };
+  }
+  // if (isRepGuild) return totalLockedResponse;
```
Comment (mprasanjith) [inline src/Modules/Guilds/Hooks/useTotalLocked.ts:71]: Should we remove this?

File: src/Modules/Guilds/Hooks/useProposal.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import { BigNumber } from 'ethers';
+
+export const formatterMiddleware = (data): Proposal => {
+  const clone = Object.assign({}, data);
+
+  //rename state to contractState
+  clone.contractState = clone.state;
+  delete clone.state;
+
+  switch (clone.contractState) {
```
Comment (mprasanjith) [inline src/Modules/Guilds/Hooks/useProposal.ts:14]: Moving this into a key-value map would be a bit cleaner.

File: src/Modules/Guilds/Hooks/useGuildConfig.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        functionName: 'getTokenVault', // Get the address of the token vault
+      },
+      {
+        ...erc20GuildContract,
+        functionName: 'getLockTime', // Get the lockTime (seconds)
+      },
+    ],
+  });
+  const { data: token } = useGuildToken(guildAddress);
+  // TODO: Move this into a SWR middleware
```
Comment (mprasanjith) [inline src/Modules/Guilds/Hooks/useGuildConfig.ts:66]: I guess this TODO is no longer valid.

**Review by mprasanjith — CHANGES_REQUESTED** [review]
Left some minor comments, but looks great! I think we can still improve TS types of some of the hooks, but that's for a separate PR.

File: src/hooks/Guilds/useUpdateEnsContent.ts (hunk 1 above)
Comment (Kenny-Gin1) [reply src/hooks/Guilds/useUpdateEnsContent.ts:5]: I thought it was the standard to write comments for `useUpdateEnsContent` since the properties are not immediately apparent. But maybe it's better to just stick to TypeScript tbh.

File: src/Modules/Guilds/Hooks/useGuildRegistry.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  const config = useNetworkConfig();
+  const { data, ...rest } = useContractRead({
+    addressOrName: contractAddress || config?.contracts?.utils.guildRegistry,
+    contractInterface: GuildRegistry.abi,
+    functionName: 'getGuildsAddresses',
+  });
+  return {
+    data,
+    ...rest,
+  };
```
Comment (MiltonTulli) [inline src/Modules/Guilds/Hooks/useGuildRegistry.ts:15]:
```suggestion
  return useContractRead({
    addressOrName: contractAddress || config?.contracts?.utils.guildRegistry,
    contractInterface: GuildRegistry.abi,
    functionName: 'getGuildsAddresses',
  });

```

File: src/Modules/Guilds/Hooks/useProposal.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  let timeDifference = moment.duration(differenceInMilliseconds).humanize();
+  if (clone.endTime.isBefore(currentTime)) {
+    clone.timeDetail = `ended ${timeDifference} ago`;
+  } else {
+    clone.timeDetail = `${timeDifference} left`;
+  }
+  return clone as Proposal;
+};
+
+export const useProposal = (guildId: string, proposalId: string) => {
```
Comment (MiltonTulli) [inline src/Modules/Guilds/Hooks/useProposal.ts:38]: All other hooks you created are exporting default. Can you refactor this too?

File: src/Modules/Guilds/Hooks/useVoterLockTimestamp.test.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import { useVoterLockTimestamp } from './useVoterLockTimestamp';
+import {
+  MOCK_CONTRACT_ADDRESS,
+  MOCK_USER_ADDRESS,
+  MOCK_BIG_NUMBER,
+} from './fixtures';
+
+jest.mock('./useVoterLockTimestamp', () => {
```
Comment (MiltonTulli) [inline src/Modules/Guilds/Hooks/useVoterLockTimestamp.test.ts:8]: Can you use absolute path for both import and mock?

File: src/components/ActionsBuilder/SupportedActions/RepMint/RepMintEditor.tsx [hunk 10]
```diff
@@ -31,29 +31,26 @@ export const Mint: React.FC<ActionEditorProps> = ({
...
+    if (data?.amount) {
+      setRepAmount((data?.amount).toString());
       setValue(
         'repPercent',
-        String((Number(initialRepAmount) * 100) / totalSupply || 0)
+        String((Number(data?.amount) * 100) / totalSupply || 0)
       );
     }
-  }, []); //eslint-disable-line
+  }, [data]); //eslint-disable-line
```
Comment (MiltonTulli) [inline src/components/ActionsBuilder/SupportedActions/RepMint/RepMintEditor.tsx:53]: If `data.amount` is primitive, for sure it will be better to put only`[data.amount]` into the deeps array. Feel free to omit this

File: src/contexts/Guilds/filters.tsx [hunk 11]
```diff
@@ -110,6 +110,7 @@ export const FilterProvider = ({
   const withFilters =
     (Component: JSX.Element) =>
     (proposal: Proposal, proposalState: ProposalState, options: Option[]) => {
+      if (!options || options.includes(undefined)) return null;
```
Comment (MiltonTulli) [inline src/contexts/Guilds/filters.tsx:113]: I don't think we should do this. If at some point options is undefined (options still loading) we want to show proposals. If we do this nothing will be returned. (my opinion)
I think you may found an issue but is regarding the next line:
`const actions = options?.map(option => option?.decodedActions).flat();`

**Review by MiltonTulli — COMMENTED** [review]
Great work @Kenny-Gin1. Looks good! 
I added some comments only regarding imports and code style.

File: src/Modules/Guilds/Hooks/useCurrentSnapshotId.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import useGuildImplementationTypeConfig from 'Modules/Guilds/Hooks/useGuildImplementationType';
```
[the remark this answers is not in the document]
Comment (Kenny-Gin1) [reply src/Modules/Guilds/Hooks/useCurrentSnapshotId.ts:1]: Tbh, the `hooks/Guilds/ should be considered deprecated. I'd rather see the structure such as `hooks/contracts`, `hooks/ens`, `hooks/erc20`.

It will be a nasty diff though. So maybe do it in the next PR?

File: src/contexts/Guilds/filters.tsx (hunk 11 above)
Comment (Kenny-Gin1) [reply src/contexts/Guilds/filters.tsx:113]: Alright! Thanks for noticing this!

**Comment by dcrescimbeni** [comment]
We have some hooks with default export, and others just individual exports. Should we at some point unify the criteria, or is it more work than what's worth?

I don't suggest doing this in this PR, just making a comment.

**Comment by VanGrim1** [comment]
> We have some hooks with default export, and others just individual exports. Should we at some point unify the criteria, or is it more work than what's worth?
> 
> I don't suggest doing this in this PR, just making a comment.

@MiltonTulli commented on it too. I took a look at the `Modules/Guilds/hooks` at least for this PR!

File: src/components/ActionsBuilder/SupportedActions/RepMint/RepMintEditor.tsx [hunk 13]
```diff
@@ -31,29 +31,26 @@ export const Mint: React.FC<ActionEditorProps> = ({
...
     context: { t },
   });
 
   useEffect(() => {
-    const initialRepAmount = ethers.utils.formatEther(
-      parsedData?.amount?.toString()
-    );
-    if (initialRepAmount) {
-      setRepAmount(initialRepAmount);
+    if (data?.amount) {
```
[the remark this answers is not in the document]
Comment (Kenny-Gin1) [reply src/components/ActionsBuilder/SupportedActions/RepMint/RepMintEditor.tsx:46]: Passed me by while refactoring. Thanks for this!

File: src/hooks/Guilds/useUpdateEnsContent.ts (hunk 1 above)
Comment (Kenny-Gin1) [reply src/hooks/Guilds/useUpdateEnsContent.ts:5]: Had a discussion on Keybase and decided to keep the comments as-is. Going forward we can start to set a comment standard around `JSDoc?, if the rest of the squad agrees.

**Comment by Kenny-Gin1** [comment]
Hey @mprasanjith @MiltonTulli @dcrescimbeni! Think I've reviewed all your comments now. Please take a look again and see if this PR is good enough for merging! Thanks for the review work so far!

**Review by dcrescimbeni — DISMISSED** [review]
LGTM :100:

**Comment by Filipv95** [comment]
Hey @Kenny-Gin1, was doing regression testing again on this branch, noticed some issue. Basically I've created two proposals that immediatelly went to 'Rejected' state, while timer is telling that they are in fact active, pls check screenrecord:

https://user-images.githubusercontent.com/37701286/188951265-681ee7ef-1d1e-4be0-af5b-340821773680.mp4

**Review by mprasanjith — DISMISSED** [review]

**Comment by Kenny-Gin1** [comment]
> Hey @Kenny-Gin1, was doing regression testing again on this branch, noticed some issue. Basically I've created two proposals that immediatelly went to 'Rejected' state, while timer is telling that they are in fact active, pls check screenrecord:
> 
>  rejectedprop.mp4

Hey! Yeah, I made a quick fix. Got the wrong indexing and should work now. 
Screenshot 2022-09-07 at 17 40 29

**Review by mprasanjith — APPROVED** [review]
